stdlib/src/builtin/_hash.mojo

```diff
@@ -0,0 +1,179 @@
+            0x082E_FA98_EC4E_6C89,
+        )
+        self.buffer = pi_key[0]
+        self.pad = pi_key[1]
+        self.extra_keys = U128(pi_key[2], pi_key[3])
+
+    @always_inline
+    fn large_update(inout self, new_data: U128):
+        """Update the buffer value with new data.
+
+        Args:
+            new_data: Value used for update.
+        """
+        var xored = new_data ^ self.extra_keys
+        var combined = _folded_multiply(xored[0], xored[1])
+        self.buffer = rotate_bits_left[ROT]((self.buffer + self.pad) ^ combined)
+
+    @always_inline
+    fn finish(self) -> UInt64:
+        """Computes the hash value based on all the previously provided data.
+
+        Returns:
+            Final hash value.
+        """
+        var rot = self.buffer & 63
+        var folded = _folded_multiply(self.buffer, self.pad)
+        return (folded << rot) | (folded >> (64 - rot))
+
+    @always_inline
+    fn write(inout self, data: UnsafePointer[UInt8], length: Int):
+        """Consume provided data to update the internal buffer.
+
+        Args:
+            data: Pointer to the byte array.
+            length: The length of the byte array.
+        """
+        self.buffer = (self.buffer + length) * MULTIPLE
+        if length > 8:
+            if length > 16:
+                var tail = data.offset(length - 16).bitcast[
+                    DType.uint64
+                ]().load[width=2]()
+                self.large_update(tail)
+                var offset = 0
+                while length - offset > 16:
+                    var block = data.offset(offset).bitcast[
+                        DType.uint64
+                    ]().load[width=2]()
+                    self.large_update(block)
+                    offset += 16
+            else:
+                var a = data.bitcast[DType.uint64]().load()
+                var b = data.offset(length - 8).bitcast[DType.uint64]().load()
+                self.large_update(U128(a, b))
+        else:
+            var value = read_small(data, length)
+            self.large_update(value)
+
+
+fn hash[
+    key: U256 = U256(0, 0, 0, 0)
+](bytes: UnsafePointer[UInt8], n: Int) -> UInt:
+    """Hash a byte array using an adopted AHash algorithm.
+
+    References:
+
+    - [Reference Implementation in Rust](https://github.com/tkaitchuck/aHash)
+
+    ```mojo
+    from random import rand
+    var n = 64
+    var rand_bytes = UnsafePointer[UInt8].alloc(n)
+    rand(rand_bytes, n)
+    _ = hash(rand_bytes, n)
+    ```
+
+    Parameters:
+        key: A key to modify the result of the hash function, defaults to [0, 0, 0, 0].
+
+    Args:
+        bytes: The byte array to hash.
+        n: The length of the byte array.
+
+    Returns:
+        A 64-bit integer hash. This hash is _not_ suitable for
+        cryptographic purposes, but will have good low-bit
+        hash collision statistical properties for common data structures.
+    """
+
+    var hasher = AHasher(key)
+    hasher.write(bytes, n)
```

Comment:
NIT:
```suggestion
    hasher.update(bytes, n)
```

Reply:
+1 to `update` per https://github.com/modularml/mojo/blob/nightly/proposals/improved-hash-module.md#introduce-a-hasher-trait anyways.  This is all subject to change as we don't have the trait yet, so just something to consider.  What do you think, @mzaks?